packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -48,17 +48,35 @@ export interface CompiledBlueprint {
 	run: (playground: UniversalPHP) => Promise<void>;
 }
 
+export type OnStepProgress = (
+	promise: Promise<void>,
+	step: StepDefinition
+) => void;
 export type OnStepCompleted = (output: any, step: StepDefinition) => any;
+export type OnStepsCompleted = (
+	completedStatus: Map<StepDefinition, boolean>
+) => void;
 
 export interface CompileBlueprintOptions {
 	/** Optional progress tracker to monitor progress */
 	progress?: ProgressTracker;
 	/** Optional semaphore to control access to a shared resource */
 	semaphore?: Semaphore;
 	/** Optional callback with step output */
+	onStepProgress?: OnStepProgress;
 	onStepCompleted?: OnStepCompleted;
+	onStepsCompleted?: OnStepsCompleted;
 }
 
+/**
+ * Convert a Promise into a boolean indicating if it resolved or rejeted.
+ */
+const didResolve = <P extends Promise<any>>(promise: P): Promise<boolean> =>
```

Comment:
`didResolve() === false` suggests the promise may still be pending. I wonder what would be a better name here.

Reply:
It was tough to find a suitable naming for this `didResolve` function. I consulted [MDN: Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) to see what the standard terminology is.

> A Promise is in one of these states:
>
> - **pending**: initial state, neither fulfilled nor rejected.
> - **fulfilled**: meaning that the operation was completed successfully.
> - **rejected**: meaning that the operation failed.
>
> A promise is said to be **settled** if it is either fulfilled or rejected, but not pending.

In the end I settled on a verbose name `isSettledPromiseResolved`, to make it clear that it's checking an already settled promise for its resolved status.